enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/impl/service/IdentityVerificationPrecompleteGuard.java

```diff
@@ -0,0 +1,173 @@
+import static com.wultra.app.enrollmentserver.model.enumeration.IdentityVerificationStatus.ACCEPTED;
+import static com.wultra.app.enrollmentserver.model.enumeration.IdentityVerificationStatus.VERIFICATION_PENDING;
+
+/**
+ * Validate all critical conditions were met before finishing the onboarding.
+ * <p>
+ * This should never happen for the state machine.
+ * It works as a safety stop.
+ *
+ * @author Lubos Racansky, lubos.racansky@wultra.com
+ */
+// TODO (racansky, 2022-10-14) consider make it Guard for Spring State Machine
+@Component
+@Slf4j
+class IdentityVerificationPrecompleteGuard {
+
+    private final IdentityVerificationConfig identityVerificationConfig;
+
+    private final RequiredDocumentTypesGuard requiredDocumentTypesGuard;
+
+    private final OnboardingOtpRepository onboardingOtpRepository;
+
+    private final ActivationService activationService;
+
+    IdentityVerificationPrecompleteGuard(
+            final IdentityVerificationConfig identityVerificationConfig,
+            final RequiredDocumentTypesGuard requiredDocumentTypesGuard,
+            final OnboardingOtpRepository onboardingOtpRepository,
+            final ActivationService activationService) {
+        this.identityVerificationConfig = identityVerificationConfig;
+        this.requiredDocumentTypesGuard = requiredDocumentTypesGuard;
+        this.onboardingOtpRepository = onboardingOtpRepository;
+        this.activationService = activationService;
+    }
+
+    /**
+     * Evaluate all precomplete conditions.
+     *
+     * @param idVerification identity verification to evaluate
+     * @return evaluation result
+     */
+    Result evaluate(final IdentityVerificationEntity idVerification) throws RemoteCommunicationException {
+        final List<DocumentVerificationEntity> documentVerifications = idVerification.getDocumentVerifications().stream()
+                .filter(it -> DocumentStatus.ALL_PROCESSED.contains(it.getStatus()))
+                .collect(Collectors.toList());
+
+        if (!documentVerifications.stream()
+                .map(DocumentVerificationEntity::getStatus)
+                .allMatch(it -> it == DocumentStatus.ACCEPTED)) {
+            logger.debug("Some documents are not accepted for identity verification ID: {}", idVerification.getProcessId());
+            return Result.failed("Some documents not accepted");
+        }
+
+        if (!requiredDocumentTypesGuard.evaluate(documentVerifications, idVerification.getId())) {
+            logger.debug("Not all required documents are present for verification ID: {}", idVerification.getProcessId());
+            return Result.failed("Required documents not present");
+        }
+
+        if (!isPrefinalPhaseAndStateValid(idVerification)) {
+            logger.debug("Not valid phase and state for verification ID: {}", idVerification.getProcessId());
+            return Result.failed("Not valid phase and state");
+        }
+
+        if (!isOtpValid(idVerification)) {
```

Comment:
It would be nice to validate both activation and user verification OTPs. I assume it's not too difficult, can we add this?